lib/Dialect/Seq/Transforms/SinkClockGates.cpp

```diff
@@ -0,0 +1,348 @@
+  /// Add the given values as operands to the `instance`. Add constant true op
+  /// as input to the other instances of the module.
+  void appendInputsToModuleAndInstances(
+      HWInstanceLike instance, ArrayRef<HWInstanceLike> moduleInstances,
+      ArrayRef<std::pair<StringAttr, Value>> inputs, HWModuleOp refMod) {
+
+    if (inputs.empty())
+      return;
+    SmallVector<Attribute> argNames;
+    if (auto args = instance->getAttrOfType<ArrayAttr>("argNames"))
+      argNames.insert(argNames.begin(), args.begin(), args.end());
+
+    SmallVector<Value> newOperands;
+    newOperands.reserve(inputs.size());
+    SmallVector<std::tuple<StringAttr, Type, Location>> newInputs;
+    newInputs.reserve(inputs.size());
+    for (const auto &[name, value] : inputs) {
+      // If no argNames attribute, donot add it.
+      if (!argNames.empty())
+        argNames.push_back(name);
+      // Additional operands to the Instance op.
+      newOperands.push_back(value);
+      // Prepare the additional input ports that need to be added to the
+      // referenced module.
+      newInputs.emplace_back(name, value.getType(), value.getLoc());
+    }
+    // Add the new operands to the instance.
+    instance->insertOperands(instance->getNumOperands(), newOperands);
+    if (!argNames.empty())
+      instance->setAttr("argNames",
+                        ArrayAttr::get(instance->getContext(), argNames));
+
+    // Now update other instances of the module.
+    for (auto otherInstance : moduleInstances) {
+      if (otherInstance == instance)
+        continue;
+      SmallVector<Value, 2> otherInstanceNewOperands;
+      otherInstanceNewOperands.reserve(inputs.size());
+      mlir::OpBuilder builder(otherInstance);
+      for (unsigned i = 0; i < inputs.size(); ++i)
+        otherInstanceNewOperands.emplace_back(builder.create<ConstantOp>(
+            otherInstance->getLoc(), builder.getI1Type(), 1));
+
+      otherInstance->insertOperands(otherInstance->getNumOperands(),
+                                    otherInstanceNewOperands);
+      if (otherInstance->hasAttr("argNames"))
+        otherInstance->setAttr(
+            "argNames", ArrayAttr::get(instance->getContext(), argNames));
+    }
+    // Update all the modules with the new input ports.
+    refMod.appendPorts(newInputs, {});
+  }
+
+  /// Get the referenced module for an instance, and add all the instances of
+  /// the module to `instances`.
+  HWModuleOp getReferencedModules(HWInstanceLike inst,
+                                  SmallVector<HWInstanceLike, 2> &instances) {
+    HWModuleOp instModule = {};
+    auto refModNames = inst.getReferencedModuleNamesAttr();
+    if (refModNames.size() != 1) {
+      inst.emitWarning("expected an instance with a single reference");
+      return {};
+    }
+    auto *node = graph->lookup(cast<StringAttr>(*refModNames.begin()));
+    if (node)
+      instModule = dyn_cast_or_null<HWModuleOp>(*node->getModule());
+    if (instModule.isPublic())
```

Comment:
Please insert null-guard for instModule. I agree that we shouldn't optimize across public modules but please add a comment and test.